src/typecheck.cpp

```diff
@@ -87,17 +86,24 @@ void typecheck(TypedAST::Identifier* ast, TypeChecker& tc) {
 	TypedAST::Declaration* declaration = ast->m_declaration;
 	assert(declaration && "accessed an unmatched identifier");
 
-	// here we implement the [var] rule
-	// TODO: refactor
-	MonoId mono = -1;
-	if (declaration->m_is_polymorphic) {
-		mono = tc.m_core.inst_fresh(declaration->m_decl_type);
+	MetaTypeId meta_type = tc.m_meta_core.find(declaration->m_meta_type);
+
+	ast->m_meta_type = meta_type;
+	if (meta_type == tc.meta_value()) {
+		// here we implement the [var] rule
+		if (declaration->m_is_polymorphic)
+			ast->m_value_type = tc.m_core.inst_fresh(declaration->m_decl_type);
+		else
+			ast->m_value_type = declaration->m_value_type;
+	} else if(meta_type == tc.meta_typefunc()){
```

Comment:
Insert a space before open brace

Reply:
will do

Reply:
Oh and also space before open parenthesis, completely did not see that